sys/net/application_layer/sock_dodtls/sock_dodtls.c

```diff
@@ -205,55 +194,14 @@ static int _connect_server(const sock_udp_ep_t *server,
         }
         goto exit;
     }
-    if ((res = sock_dtls_create(&_dtls_sock, &_udp_sock, creds->tag,
-                                SOCK_DTLS_1_2, SOCK_DTLS_CLIENT)) < 0) {
-        puts("Unable to create DTLS sock\n");
-        _close_session(creds->tag, creds->type);
-        goto exit;
-    }
-
-    start = _now_ms();
-    try_start = start;
-    while (((try_start = _now_ms()) - start) < SOCK_DODTLS_SESSION_TIMEOUT_MS) {
-        memset(&_server_session, 0, sizeof(_server_session));
-        if ((res = sock_dtls_session_init(&_dtls_sock, server,
-                                          &_server_session)) >= 0) {
-            uint32_t try_duration;
-
-            res = sock_dtls_recv(&_dtls_sock, &_server_session, _dns_buf,
-                                 sizeof(_dns_buf), timeout * US_PER_MS);
-            if (res == -SOCK_DTLS_HANDSHAKE) {
-                break;
-            }
-            DEBUG("Unable to establish DTLS handshake: %d (timeout: %luus)\n",
-                  -res, (long unsigned)timeout * US_PER_MS);
-            sock_dtls_session_destroy(&_dtls_sock, &_server_session);
-            try_duration = _now_ms() - try_start;
-            if (try_duration < timeout) {
-                _sleep_ms(timeout - try_duration);
-            }
-            /* see https://datatracker.ietf.org/doc/html/rfc6347#section-4.2.4.1 */
-            timeout *= 2U;
-        }
-        else {
-            DEBUG("Unable to initialize DTLS session: %d\n", -res);
-            sock_dtls_session_destroy(&_dtls_sock, &_server_session);
-        }
-    }
-    if (res != -SOCK_DTLS_HANDSHAKE) {
-        res = -ETIMEDOUT;
-        _close_session(creds->tag, creds->type);
-        goto exit;
-    }
-    else {
-        res = 0;
-    }
 
+    res = sock_dtls_establish_session(&_udp_sock, &_dtls_sock, &_server_session,
+                                      creds->tag, &local, server, _dns_buf,
+                                      sizeof(_dns_buf));
     _cred_type = creds->type;
     _cred_tag = creds->tag;
     _id = (uint16_t)(random_uint32() & 0xffff);
 exit:
-    memset(_dns_buf, 0, sizeof(_dns_buf));  /* flush-out unencrypted data */
     mutex_unlock(&_server_mutex);
     return (res > 0) ? 0 : res;
```

Comment:
@benpicco: the `credman_delete(creds_tag, creds_type);` got removed from line 184 maybe we need  that in l. 206 if the connection could not be established since in that case it would not be closed -> that credman slot would be taken until reboot.

Reply:
Kind of what I thought [here](https://github.com/RIOT-OS/RIOT/pull/19142#discussion_r1103866067).

Reply:
maybe that whole file should never call any credman function (but find or get) since the add should be done  in another part of RIOT

Reply:
seems more like it is not used as intended

intention:
init: 
add all your creeds to credman 

connect:

use a cred by tag 

but there where applications before that that did the credential handling themself and these are now add cred to credman
and that use the cred


if we use it as intended we would just have one place where all the creds are managed this would make the review of the creds on a device simpler)

Reply:
https://github.com/RIOT-OS/RIOT/pull/11564#issuecomment-496631906

there is the reasoning why credman and not just a pointer to credentials